### PR TITLE
add block cache for upgrade service

### PR DESCRIPTION
**Pull request by mariobassem**
## Related issues:

- #2000

### CODE REVIEW

File: pkg/upgrade/cache.go [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+	"golang.org/x/crypto/blake2b"
+
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	// DefaultDownloadWorkers define the default number of workload to use to downloads data blocks
+	DefaultDownloadWorkers = 4
+	//DefaultBlockSize is the default block size
+	DefaultBlockSize = 512 //KB
```
Comment (xmonader) [inline pkg/upgrade/cache.go]: Please add unit suffix e.g DefaultBlockSizeKB

Comment (xmonader) [reply pkg/upgrade/cache.go]: also noticed these constants are also used internally in functions and there's no way to override them, please make them unexported

File: pkg/upgrade/cache.go [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+)
+
+// Downloader allows to get some data blocks using a pool of workers
+type Downloader struct {
+	cache     string
```
Comment (xmonader) [inline pkg/upgrade/cache.go]: is that cacheDir?

File: pkg/upgrade/cache.go [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+	workers   int
```
Comment (xmonader) [inline pkg/upgrade/cache.go]: change to numWorkers

File: pkg/upgrade/cache.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+	storage   storage.Storage
+	blocks    []meta.BlockInfo
+	blockSize uint64
```
Comment (xmonader) [inline pkg/upgrade/cache.go]: add unit here please, Is it byte, kb, .. etc

File: pkg/upgrade/cache.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+)
+
+const (
+	// DefaultDownloadWorkers define the default number of workload to use to downloads data blocks
+	DefaultDownloadWorkers = 4
+	//DefaultBlockSize is the default block size
+	DefaultBlockSize = 512 //KB
+)
+
+// Downloader allows to get some data blocks using a pool of workers
```
Comment (xmonader) [inline pkg/upgrade/cache.go]: This is a concrete type, I believe the documentation needs to be more specific

File: pkg/upgrade/cache.go [hunk 6]
```diff
@@ -43,11 +43,11 @@ type DownloaderOption interface {
 	apply(d *Downloader)
```
Comment (xmonader) [inline pkg/upgrade/cache.go]: it's not done like that, please make sure to remove the interface, 

you can have a new type e.g DownloaderOption that's a mutation method func(d  *Downloader)

and create function e.g WtihNumWorkers e.g

```go
func WithNumWorkers(n int) func(*Downloader) {
            return func(d *Downloader){ 
                    d.numWorkers = n
            
            }
}

```


Please don't introduce interfaces that're not needed

**Review by xmonader — CHANGES_REQUESTED** [review]

File: pkg/upgrade/cache.go (hunk 6 above)
Comment (xmonader) [reply pkg/upgrade/cache.go]: we use that in order not to break the API, while this is all internal it's fine to have the numworker in the factory function, or even if the type is simple an depends on the zero value you can directly use it and have the fields exported without tons of chores.

**Review by muhamadazmy — DISMISSED** [review]

**Review by xmonader — APPROVED** [review]

**Pull request merged**